### PR TITLE
Refactor: Felt252 use lambdaworks field element

### PR DESCRIPTION
**Pull request by pefontana**
# TITLE

## Description

Closes: #1197

This PR changes current `Felt252` backend from _num_bigint_'s `FeltBigInt` to _lambdawork_'s `FieldElement<Stark252PrimeField>`.

# Step by Step to benchmark lambdawork FieldElement:

## Create binaries
```
mkdir binaries
```

###  Create binary from main 
// Main las commit
```
git checkout de6a2327b7e629336bc2431787c8e28c3b0f1349 
```

```
cargo build -p cairo-vm-cli --release
```

```
mv target/release/cairo-vm-cli binaries/cairo-vm-cli-main
```

###  Create binary from lambdaworks-felt 

```
cargo clean
```

// lambdaworks-felt last commit (cambiarlo con el ultimo commit)
```
git checkout 7dd7d58cabdb4861549149431db0461d965ec531 
```

```
cargo build -p cairo-vm-cli --release
```

```
mv target/release/cairo-vm-cli binaries/cairo-vm-cli-lambdaworks
```

## Compile cairo program
Install dependencies if neesary

Linux users
```
make deps
```

Mac users
```
make deps-macos
```

```
source cairo-rs-env/bin/activate
```

```
make cairo_bench_programs
```

## Run benchmarks
```
chmod +x lamdaworks-benchmarks.sh
```

```
sh lamdaworks-benchmarks.sh
```

### CODE REVIEW

**Comment by lambda-0x** [comment]
Benchmark ran on x86_64 machine:

image
image

File: felt/src/arbitrary.rs [hunk 1]
```diff
@@ -1,52 +1,41 @@
...
-use crate::{
-    bigint_felt::{FeltBigInt, FIELD_HIGH, FIELD_LOW},
-    Felt252,
-};
+use crate::{Felt252, FIELD_HIGH, FIELD_LOW};
 
-fn any_felt_big_int<const PH: u128, const PL: u128>() -> impl Strategy<Value = FeltBigInt<PH, PL>> {
-    (0..=PH)
+/// Returns a [`Strategy`] that generates any valid Felt252
+fn any_felt252() -> impl Strategy<Value = Felt252> {
```
Comment (Oppen) [inline felt/src/arbitrary.rs:8]: At this point we could think about moving this to lambdaworks.

File: src/lib.rs [hunk 2]
```diff
@@ -9,6 +9,8 @@
 #![deny(warnings)]
 #![forbid(unsafe_code)]
 #![cfg_attr(any(target_arch = "wasm32", not(feature = "std")), no_std)]
+// TODO: remove when boxed
+#![allow(clippy::result_large_err)]
```
Comment (Oppen) [inline src/lib.rs:13]: I think we did box already, so this can go, right?

File: felt/src/lib.rs [hunk 3]
```diff
@@ -108,162 +58,274 @@ macro_rules! felt_str {
...
-#[derive(Eq, Hash, PartialEq, PartialOrd, Ord, Clone, Deserialize, Default, Serialize)]
+impl fmt::Display for ParseFeltError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "{ParseFeltError:?}")
+    }
+}
+
+#[derive(Eq, Hash, PartialEq, Clone, Deserialize, Serialize)]
+#[serde(from = "BigInt")]
+#[serde(into = "BigInt")]
```
Comment (Oppen) [inline felt/src/lib.rs:69]: TIL. I think we can take advantage of this in program deserialization.

File: felt/src/lib.rs [hunk 4]
```diff
@@ -108,162 +58,274 @@ macro_rules! felt_str {
...
-        );
-        res[..8].copy_from_slice(&d0);
-        res[8..16].copy_from_slice(&d1);
-        res[16..24].copy_from_slice(&d2);
-        res[24..].copy_from_slice(&d3);
-        res
+        // TODO: upstream should return array
+        self.value
+            .representative()
+            .to_bytes_le()
```
Comment (Oppen) [inline felt/src/lib.rs:210]: This makes me worry a bit in terms of performance due to the allocation. Can't we iter the digits of the representative as it was done originally?

File: felt/src/lib.rs [hunk 5]
```diff
@@ -108,162 +58,274 @@ macro_rules! felt_str {
...
-        let mut iter = self.iter_u64_digits();
-        [
-            iter.next().unwrap_or_default(),
-            iter.next().unwrap_or_default(),
-            iter.next().unwrap_or_default(),
-            iter.next().unwrap_or_default(),
-        ]
+        let mut rep = self.value.representative();
+        rep.limbs.reverse();
+        rep.limbs
```
Comment (Oppen) [inline felt/src/lib.rs:227]: Just like we do here!

File: felt/src/lib.rs [hunk 6]
```diff
@@ -343,31 +409,27 @@ impl Add<&Felt252> for u64 {
...
+            // No digits means `rhs` is `0`, so the sum is simply `self`.
+            [0, 0, 0, 0] => Some(self),
+            // A single digit means this is effectively the sum of two `u64` numbers.
+            [0, 0, 0, low] => self.checked_add(low),
+            // Now we need to compare the 3 most significant digits.
+            // There are two relevant cases from now on, either `rhs` behaves like a
+            // substraction of a `u64` or the result of the sum falls out of range.
+
             // The 3 MSB only match the prime for Felt252::max_value(), which is -1
             // in the signed field, so this is equivalent to substracting 1 to `self`.
```
Comment (Oppen) [inline felt/src/lib.rs:429]: The comments become misleading now :(
That description counting digits and all was quite specific to the iterative implementation. If doing digits directly it doesn't change that much.
In any case, we need to revalidate the premise that we need a specialized addition here with the new type. With `BigInt` was quite slow and a major bottleneck, but that's not necessarily still the case with `FieldElement`.

File: felt/src/lib.rs [hunk 7]
```diff
@@ -397,7 +459,8 @@ impl AddAssign for Felt252 {
 
 impl<'a> AddAssign<&'a Felt252> for Felt252 {
     fn add_assign(&mut self, rhs: &Self) {
-        self.value += &rhs.value;
+        // TODO: optimize and move upstream
```
Comment (Oppen) [inline felt/src/lib.rs:462]: I think they have a `+=` now. It's not optimized tho, it's pretty much the same implementation.

File: felt/src/lib.rs [hunk 8]
```diff
@@ -455,51 +518,52 @@ impl<'a> Sub<&'a Felt252> for Felt252 {
...
 impl Sub<&Felt252> for usize {
     type Output = Felt252;
     fn sub(self, rhs: &Self::Output) -> Self::Output {
-        Self::Output {
-            value: self - &rhs.value,
-        }
+        let neg = Self::Output {
+            value: (&rhs.value).neg(),
+        };
+        neg + self
```
Comment (Oppen) [inline felt/src/lib.rs:529]: Is there no substraction upstream?

File: src/lib.rs (hunk 2 above)
Comment (MegaRedHand) [reply src/lib.rs:13]: Yes!

File: felt/src/arbitrary.rs (hunk 1 above)
Comment (MegaRedHand) [reply felt/src/arbitrary.rs:8]: Yes, thanks for the suggestion! Created an issue: https://github.com/lambdaclass/lambdaworks/issues/414

File: felt/src/lib.rs (hunk 7 above)
Comment (MegaRedHand) [reply felt/src/lib.rs:462]: They have it only for owned values (i.e. for `a += b`, where both are not references) 😢

File: felt/src/lib.rs (hunk 4 above)
Comment (MegaRedHand) [reply felt/src/lib.rs]: Good idea!

**Comment by pefontana** [comment]
Benchmark ran on:
lambda server
todo: hardware specifications
todo: create table with results

main: `fc09e20b786e87d1cd341e05f491c358d032015e`
lambdaworks: `0be0200db96a4fc528d8d717ef9a55d5225ca38a`

warmups: 50
runs: 100

Running blake2s_integration_benchmark benchmark
Benchmark 1: main
  Time (mean ± σ):      1.076 s ±  0.016 s    [User: 1.025 s, System: 0.050 s]
  Range (min … max):    1.035 s …  1.104 s    100 runs

Benchmark 2: lambdaworks
  Time (mean ± σ):     779.8 ms ±  10.1 ms    [User: 734.2 ms, System: 44.9 ms]
  Range (min … max):   762.4 ms … 839.4 ms    100 runs

Summary
  'lambdaworks' ran
    1.38 ± 0.03 times faster than 'main'
Running compare_arrays_200000 benchmark
Benchmark 1: main
  Time (mean ± σ):      1.245 s ±  0.013 s    [User: 1.170 s, System: 0.074 s]
  Range (min … max):    1.182 s …  1.269 s    100 runs

Benchmark 2: lambdaworks
  Time (mean ± σ):     792.3 ms ±   7.5 ms    [User: 716.0 ms, System: 75.7 ms]
  Range (min … max):   771.0 ms … 824.2 ms    100 runs

Summary
  'lambdaworks' ran
    1.57 ± 0.02 times faster than 'main'
Running dict_integration_benchmark benchmark
Benchmark 1: main
  Time (mean ± σ):      1.010 s ±  0.011 s    [User: 0.966 s, System: 0.043 s]
  Range (min … max):    0.993 s …  1.040 s    100 runs

Benchmark 2: lambdaworks
  Time (mean ± σ):     683.7 ms ±   7.8 ms    [User: 638.1 ms, System: 45.0 ms]
  Range (min … max):   667.8 ms … 716.2 ms    100 runs

Summary
  'lambdaworks' ran
    1.48 ± 0.02 times faster than 'main'
Running factorial_multirun benchmark
Benchmark 1: main
  Time (mean ± σ):      1.480 s ±  0.016 s    [User: 1.405 s, System: 0.074 s]
  Range (min … max):    1.425 s …  1.517 s    100 runs

Benchmark 2: lambdaworks
  Time (mean ± σ):     741.9 ms ±  18.4 ms    [User: 662.3 ms, System: 78.9 ms]
  Range (min … max):   727.7 ms … 859.1 ms    100 runs

  Warning: Statistical outliers were detected. Consider re-running this benchmark on a quiet system without any interferences from other programs. It might help to use the '--warmup' or '--prepare' options.

Summary
  'lambdaworks' ran
    2.00 ± 0.05 times faster than 'main'
Running fibonacci_1000_multirun benchmark
Benchmark 1: main
  Time (mean ± σ):      1.135 s ±  0.012 s    [User: 1.061 s, System: 0.074 s]
  Range (min … max):    1.127 s …  1.229 s    100 runs

  Warning: Statistical outliers were detected. Consider re-running this benchmark on a quiet system without any interferences from other programs. It might help to use the '--warmup' or '--prepare' options.

Benchmark 2: lambdaworks
  Time (mean ± σ):     701.0 ms ±  11.3 ms    [User: 625.8 ms, System: 74.6 ms]
  Range (min … max):   668.5 ms … 757.3 ms    100 runs

Summary
  'lambdaworks' ran
    1.62 ± 0.03 times faster than 'main'
Running field_arithmetic_get_square_benchmark benchmark
Benchmark 1: main
  Time (mean ± σ):     472.9 ms ±   3.8 ms    [User: 452.6 ms, System: 19.8 ms]
  Range (min … max):   451.1 ms … 480.2 ms    100 runs

Benchmark 2: lambdaworks
  Time (mean ± σ):     389.2 ms ±   7.3 ms    [User: 364.2 ms, System: 24.6 ms]
  Range (min … max):   383.6 ms … 445.5 ms    100 runs

  Warning: Statistical outliers were detected. Consider re-running this benchmark on a quiet system without any interferences from other programs. It might help to use the '--warmup' or '--prepare' options.

Summary
  'lambdaworks' ran
    1.22 ± 0.02 times faster than 'main'
Running integration_builtins benchmark
Benchmark 1: main
  Time (mean ± σ):      1.205 s ±  0.012 s    [User: 1.133 s, System: 0.071 s]
  Range (min … max):    1.174 s …  1.303 s    100 runs

  Warning: Statistical outliers were detected. Consider re-running this benchmark on a quiet system without any interferences from other programs. It might help to use the '--warmup' or '--prepare' options.

Benchmark 2: lambdaworks
  Time (mean ± σ):     988.4 ms ±   9.0 ms    [User: 908.1 ms, System: 79.3 ms]
  Range (min … max):   970.5 ms … 1040.7 ms    100 runs

  Warning: Statistical outliers were detected. Consider re-running this benchmark on a quiet system without any interferences from other programs. It might help to use the '--warmup' or '--prepare' options.

Summary
  'lambdaworks' ran
    1.22 ± 0.02 times faster than 'main'
Running keccak_integration_benchmark benchmark
Benchmark 1: main
  Time (mean ± σ):      1.176 s ±  0.007 s    [User: 1.108 s, System: 0.068 s]
  Range (min … max):    1.156 s …  1.203 s    100 runs

Benchmark 2: lambdaworks
  Time (mean ± σ):     964.1 ms ±  11.5 ms    [User: 902.1 ms, System: 61.2 ms]
  Range (min … max):   948.6 ms … 1013.1 ms    100 runs

Summary
  'lambdaworks' ran
    1.22 ± 0.02 times faster than 'main'
Running linear_search benchmark
Benchmark 1: main
  Time (mean ± σ):      1.301 s ±  0.015 s    [User: 1.229 s, System: 0.071 s]
  Range (min … max):    1.262 s …  1.352 s    100 runs

Benchmark 2: lambdaworks
  Time (mean ± σ):     838.1 ms ±  11.3 ms    [User: 760.3 ms, System: 77.1 ms]
  Range (min … max):   821.3 ms … 901.4 ms    100 runs

  Warning: Statistical outliers were detected. Consider re-running this benchmark on a quiet system without any interferences from other programs. It might help to use the '--warmup' or '--prepare' options.

Summary
  'lambdaworks' ran
    1.55 ± 0.03 times faster than 'main'
Running math_cmp_and_pow_integration_benchmark benchmark
Benchmark 1: main
  Time (mean ± σ):      1.023 s ±  0.004 s    [User: 0.951 s, System: 0.071 s]
  Range (min … max):    1.015 s …  1.041 s    100 runs

Benchmark 2: lambdaworks
  Time (mean ± σ):     722.7 ms ±   8.7 ms    [User: 639.5 ms, System: 82.5 ms]
  Range (min … max):   705.6 ms … 771.1 ms    100 runs

Summary
  'lambdaworks' ran
    1.42 ± 0.02 times faster than 'main'
Running math_integration_benchmark benchmark
Benchmark 1: main
  Time (mean ± σ):     984.4 ms ±   8.7 ms    [User: 942.1 ms, System: 41.6 ms]
  Range (min … max):   952.4 ms … 1009.8 ms    100 runs

Benchmark 2: lambdaworks
  Time (mean ± σ):     706.9 ms ±  13.1 ms    [User: 656.6 ms, System: 49.6 ms]
  Range (min … max):   689.8 ms … 775.9 ms    100 runs

  Warning: Statistical outliers were detected. Consider re-running this benchmark on a quiet system without any interferences from other programs. It might help to use the '--warmup' or '--prepare' options.

Summary
  'lambdaworks' ran
    1.39 ± 0.03 times faster than 'main'
Running memory_integration_benchmark benchmark
Benchmark 1: main
  Time (mean ± σ):     845.7 ms ±   8.7 ms    [User: 804.1 ms, System: 40.9 ms]
  Range (min … max):   833.7 ms … 874.4 ms    100 runs

Benchmark 2: lambdaworks
  Time (mean ± σ):     550.7 ms ±  10.6 ms    [User: 504.3 ms, System: 45.9 ms]
  Range (min … max):   540.9 ms … 610.5 ms    100 runs

Summary
  'lambdaworks' ran
    1.54 ± 0.03 times faster than 'main'
Running operations_with_data_structures_benchmarks benchmark
Benchmark 1: main
  Time (mean ± σ):     841.7 ms ±   7.7 ms    [User: 801.3 ms, System: 39.6 ms]
  Range (min … max):   832.0 ms … 887.2 ms    100 runs

Benchmark 2: lambdaworks
  Time (mean ± σ):     637.2 ms ±   8.7 ms    [User: 590.0 ms, System: 46.6 ms]
  Range (min … max):   622.9 ms … 697.4 ms    100 runs

Summary
  'lambdaworks' ran
    1.32 ± 0.02 times faster than 'main'
Running pedersen benchmark
Benchmark 1: main
  Time (mean ± σ):     376.4 ms ±   1.7 ms    [User: 364.9 ms, System: 11.1 ms]
  Range (min … max):   374.2 ms … 388.7 ms    100 runs

Benchmark 2: lambdaworks
  Time (mean ± σ):     336.9 ms ±   1.3 ms    [User: 323.2 ms, System: 13.4 ms]
  Range (min … max):   335.1 ms … 343.8 ms    100 runs

Summary
  'lambdaworks' ran
    1.12 ± 0.01 times faster than 'main'
Running poseidon_integration_benchmark benchmark
Benchmark 1: main
  Time (mean ± σ):     792.4 ms ±   6.3 ms    [User: 779.3 ms, System: 12.5 ms]
  Range (min … max):   784.7 ms … 805.1 ms    100 runs

Benchmark 2: lambdaworks
  Time (mean ± σ):     701.5 ms ±   4.6 ms    [User: 686.1 ms, System: 14.8 ms]
  Range (min … max):   693.6 ms … 714.0 ms    100 runs

Summary
  'lambdaworks' ran
    1.13 ± 0.01 times faster than 'main'
Running secp_integration_benchmark benchmark
Benchmark 1: main
  Time (mean ± σ):      1.089 s ±  0.016 s    [User: 1.056 s, System: 0.032 s]
  Range (min … max):    1.041 s …  1.113 s    100 runs

Benchmark 2: lambdaworks
  Time (mean ± σ):     813.4 ms ±  17.6 ms    [User: 781.9 ms, System: 30.9 ms]
  Range (min … max):   781.8 ms … 909.6 ms    100 runs

Summary
  'lambdaworks' ran
    1.34 ± 0.03 times faster than 'main'
Running set_integration_benchmark benchmark
Benchmark 1: main
  Time (mean ± σ):     806.9 ms ±   3.9 ms    [User: 794.1 ms, System: 12.1 ms]
  Range (min … max):   800.1 ms … 818.4 ms    100 runs

Benchmark 2: lambdaworks
  Time (mean ± σ):     456.3 ms ±   4.5 ms    [User: 444.0 ms, System: 11.9 ms]
  Range (min … max):   450.0 ms … 476.3 ms    100 runs

  Warning: Statistical outliers were detected. Consider re-running this benchmark on a quiet system without any interferences from other programs. It might help to use the '--warmup' or '--prepare' options.

Summary
  'lambdaworks' ran
    1.77 ± 0.02 times faster than 'main'
Running uint256_integration_benchmark benchmark
Benchmark 1: main
  Time (mean ± σ):      1.483 s ±  0.014 s    [User: 1.408 s, System: 0.073 s]
  Range (min … max):    1.472 s …  1.615 s    100 runs

  Warning: Statistical outliers were detected. Consider re-running this benchmark on a quiet system without any interferences from other programs. It might help to use the '--warmup' or '--prepare' options.

Benchmark 2: lambdaworks
  Time (mean ± σ):      1.112 s ±  0.014 s    [User: 1.025 s, System: 0.086 s]
  Range (min … max):    1.086 s …  1.201 s    100 runs

Summary
  'lambdaworks' ran
    1.33 ± 0.02 times faster than 'main'

File: .github/workflows/hyperfine.yml [hunk 9]
```diff
@@ -156,8 +154,8 @@ jobs:
           done
         fi
         find target_programs -name '*.json' -exec basename -s .json '{}' '+' | \
-        sort | xargs -P 2 -I '{program}' \
-        hyperfine -N -r 10 --export-markdown "target_programs/{program}.md" \
+        sort | xargs -I '{program}' \
+        hyperfine -N -w 10 -r 20 --export-markdown "target_programs/{program}.md" \
```
Comment (Oppen) [inline .github/workflows/hyperfine.yml:158]: We should probably revert these to the original values so pipelines don't take too long.

File: CHANGELOG.md [hunk 10]
```diff
@@ -2,6 +2,8 @@
 
 #### Upcoming Changes
 
+* feat: `Felt252` now uses _lambdaworks_' `FieldElement` internally
```
Comment (Oppen) [inline CHANGELOG.md:5]: AFAICT it's a breaking change. We removed a `trait` among several other things.

File: src/hint_processor/builtin_hint_processor/math_utils.rs [hunk 11]
```diff
@@ -117,41 +117,49 @@ pub fn assert_le_felt(
         .ok_or_else(|| HintError::MissingConstant(Box::new(PRIME_OVER_2_HIGH)))?;
     let a = &get_integer_from_var_name("a", vm, ids_data, ap_tracking)?
         .clone()
-        .into_owned();
+        .into_owned()
+        .to_biguint();
```
Comment (Oppen) [inline src/hint_processor/builtin_hint_processor/math_utils.rs:121]: I'm not sure we need all this chain. Doesn't simply `.to_biguint()` do the work here?

File: src/hint_processor/builtin_hint_processor/math_utils.rs [hunk 12]
```diff
@@ -117,41 +117,49 @@ pub fn assert_le_felt(
...
 
+    // TODO: use UnsignedInteger for this
+    let prime_div2 = prime_div_constant(2)?;
+    let prime_div3 = prime_div_constant(3)?;
+
     if a > b {
-        return Err(HintError::NonLeFelt252(Box::new((a.clone(), b.clone()))));
+        return Err(HintError::NonLeFelt252(Box::new((
+            Felt252::from(a.clone()),
+            Felt252::from(b.clone()),
```
Comment (Oppen) [inline src/hint_processor/builtin_hint_processor/math_utils.rs:135]: Do we need to `clone` here? Won't the reference do?

**Review by Oppen — APPROVED** [review]

File: src/hint_processor/builtin_hint_processor/math_utils.rs (hunk 12 above)
Comment (MegaRedHand) [reply src/hint_processor/builtin_hint_processor/math_utils.rs:135]: We don't have a `From<&BigUint>` implementation. Although, I found a way to remove these: I was taking a reference to `a` and `b` unnecessarily, just removing that will do.

File: CHANGELOG.md (hunk 10 above)
Comment (MegaRedHand) [reply CHANGELOG.md:5]: Added the changes to the changelog. We only removed `modpow` and `to_bytes_be`. The `FeltOps` trait we removed was `pub(crate)` 🗿

File: felt/src/lib.rs [hunk 13]
```diff
@@ -678,13 +688,12 @@ impl Zero for Felt252 {
...
-        Self {
-            value: FieldElement::<Stark252PrimeField>::one(),
-        }
+        let value = FieldElement::from_raw(&Stark252PrimeField::ONE);
+        Self { value }
     }
 
     fn is_one(&self) -> bool {
-        self.value == FieldElement::<Stark252PrimeField>::one()
+        self.value == FieldElement::from_raw(&Stark252PrimeField::ONE)
```
[the remark this answers is not in the document]
Comment (MegaRedHand) [reply felt/src/lib.rs:696]: I think you are on an old version. This change was already done last week. The implementation of `FieldElement::one` uses `Stark252PrimeField::one`, which uses `Stark252PrimeField::ONE`, so it should work the same both ways.

File: CHANGELOG.md [hunk 14]
```diff
@@ -2,6 +2,9 @@
...
```
Comment (Oppen) [inline CHANGELOG.md:5]: If it implements `Copy` now, add it to the changelog.

**Pull request closed without merging**